test/opt/test_unify_const.cpp

```diff
@@ -0,0 +1,754 @@
+          // preparing for the future updates of SPIR-V.
+          // TODO(qining): change to a valid decoration once they are available.
+          "OpDecorate %s_duplicate RelaxedPrecision",
+      })
+      .AppendTypesConstantsGlobals({
+          // clang-format off
+          "%bool_true = OpConstantTrue %bool",
+          "%signed_1 = OpConstant %int 1",
+          "%float_1 = OpConstant %float 1",
+          "%double_1 = OpConstant %double 1",
+          "%s = OpConstantComposite %flat_struct %bool_true %signed_1 %float_1 %double_1",
+          // The duplicated constants without decorations should not be skipped.
+          "%float_1_duplicate = OpConstant %float 1",
+          "%double_1_duplicate = OpConstant %double 1",
+          // The type "flat_struct" has decoration, so the following
+          // instruction should not be removed.
+          "%s_duplicate = OpConstantComposite %flat_struct %bool_true %signed_1 %float_1_duplicate %double_1_duplicate",
+          // clang-format on
+      })
+      .AppendInMain({
+          "%s_var = OpVariable %_pf_flat_struct Function",
+          "OpStore %s_var %s_duplicate",
+      });
+
+  expected_builder_
+      .AppendAnnotations({
+          "OpDecorate %s_duplicate RelaxedPrecision",
+      })
+      .AppendTypesConstantsGlobals({
+          // clang-format off
+          "%bool_true = OpConstantTrue %bool",
+          "%signed_1 = OpConstant %int 1",
+          "%float_1 = OpConstant %float 1",
+          "%double_1 = OpConstant %double 1",
+          "%s = OpConstantComposite %flat_struct %bool_true %signed_1 %float_1 %double_1",
+          "%s_duplicate = OpConstantComposite %flat_struct %bool_true %signed_1 %float_1 %double_1",
+          // clang-format on
+      })
+      .AppendInMain({
+          "%s_var = OpVariable %_pf_flat_struct Function",
+          "OpStore %s_var %s_duplicate",
+      })
+      .AppendNames({
+          "OpName %float_1 \"float_1_duplicate\"",
+          "OpName %double_1 \"double_1_duplicate\"",
+      });
+  Check();
+}
+
+struct UnifyConstantTestCase {
+  // new constants that should be kept.
+  std::vector<std::string> new_consts;
+  // expected uses of the new defined constants.
+  std::vector<std::string> use_new_consts;
+  // duplicated constants of the new constants.
+  std::vector<std::string> duplicate_consts;
+  // uses of the duplicated constants, expected to be updated to use the new
+  // constants.
+  std::vector<std::string> use_duplicate_consts;
+  // The updated OpName instructions of the duplicated constants.
+  std::vector<std::string> remapped_names;
+};
+
+using UnifyFrontEndConstantParamTest = UnifyConstantTest<
+    PassTest<::testing::TestWithParam<UnifyConstantTestCase>>>;
+
+TEST_P(UnifyFrontEndConstantParamTest, TestCase) {
+  auto& tc = GetParam();
+  expected_builder_.AppendTypesConstantsGlobals(tc.new_consts)
+      .AppendInMain(tc.use_new_consts)
+      .AppendNames(tc.remapped_names);
```

Comment:
Worth a comment here saying `// With their attached debug instruction migrated to the corresponding preserved constants.`